src/common/middleware/call-api.js

```diff
@@ -0,0 +1,101 @@
+import { checkStatus, getError } from '../helpers/api';
+
+// Action key that carries API call info interpreted by this Redux middleware.
+export const CALL_API = 'CALL_API';
+
+// A Redux middleware that interprets actions with CALL_API info specified.
+// Performs the call and promises when such actions are dispatched.
+export default defaults => () => next => async action => {
+  // If action does not invoke CALL_API,
+  // ignore it
+  const settings = action[CALL_API];
+  if (typeof settings === 'undefined') {
+    return next(action);
+  }
+
+  // Validate action settings
+  validateDefaults(defaults);
+  validateSettings(settings);
+
+  const [ successType, failureType ] = settings.types;
+  const resource = defaults.endpoint + settings.path;
+  const optionsWithCsrfToken = withCsrfToken(defaults.csrfToken, settings.options);
+  const createAction = createActionWith.bind({}, action);
+
+  // If action has a type, dispatch the action next
+  if (action.type) {
+    next(createAction());
+  }
+
+  try {
+    const response = await fetch(resource, optionsWithCsrfToken);
+    await checkStatus(response);
+    const result = await response.json();
+
+    if (result.status !== 'success') {
+      throw getError(response, result);
+    }
+
+    return next(createAction({
+      type: successType,
+      response: result.payload
```

Comment:
This follows the [Redux real-world example](https://github.com/reactjs/redux/blob/master/examples/real-world/src/middleware/api.js#L117). The reason it does this is not to conflict with the payload property from [the action it extends](https://github.com/reactjs/redux/blob/master/examples/real-world/src/middleware/api.js#L107).

I'm using this to allow the payload from the base action to be passed with the success and error actions too. This is important for actions such as `REMOVE_SNAP_SUCCESS`, because it needs the repo URL from the payload to remove the snap from the store.